### DISCUSSION TITLE
Help understanding `get_data_in_units` in `pynwb` for `ElectricalSeries`

### DISCUSSION THREAD

**Discussion by rohanshah** [NeurodataWithoutBorders/helpdesk#83] [Q&A] [answered]
Hello, I hope this is the appropriate for getting help understanding some source code in `pynwb`. Happy to file an issue in that GitHub repository instead if that is more appropriate.

I was reading through some of the source code for the `TimeSeries` and `ElectricalSeries` object and was confused by the implementation of [`get_data_in_units`](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/src/pynwb/base.py#L339-L363)... hoping to clarify my understanding with this discussion post.

It seems that when `channel_conversion` is present, the `get_data_in_units` function tries to multiply the conversion factor across the time dimension instead of the electrode dimension?

For example, if I have `data` with a shape of `(100, 2)` (i.e. `100` samples across two electrodes) and a `channel_conversion = channel_conversion = np.asarray([2.0, 3.0])` (i.e. a scaling factor of `2`  and `3` for each electrode, respectively), I get the error
```
ValueError: operands could not be broadcast together with shapes (100,2) (2,1) 
```

If I change the source code to perform this calculation then it seems to correctly scale the `data` (e.g. all samples for the first electrode multiplied by a factor of `2` and all samples for the second electrode multiplied by a factor of `3`)
```
            scale_factor = self.conversion * self.channel_conversion
```

Further adding to the confusion is the unit test for this case:

https://github.com/NeurodataWithoutBorders/pynwb/blob/90e60484415d38d3d566dc9fea63b6d99ffdbe43/tests/unit/test_ecephys.py#L119-L135

Specifically, the first few lines of setup:
```
        data = np.asarray([[1, 1, 1, 1, 1], [1, 1, 1, 1, 1]])
        conversion = 1.0
        offset = 3.0
        channel_conversion = np.asarray([2.0, 2.0])
```

It was my understanding of the shape of `data` (i.e. `(2, 5)`) that it is meant to represent 5 electrodes with 2 samples each but the `channel_conversion` is for 2 channels...

Maybe I am misunderstanding the way `data` or `channel_conversion` are meant to be shaped?

Thank you for your help!

**Comment by bendichter** [comment, marked as answer]
@rohanshah your analysis is spot on! Would you be interested in filing a pull request to fix this bug?

**Reply by rohanshah** [reply]
@bendichter absolutely! I'll work on getting a pull request opened on `pynwb` early this week. Thanks for the quick feedback on my question!

**Reply by rohanshah** [reply]
Hey @bendichter I added my fix here: https://github.com/NeurodataWithoutBorders/pynwb/pull/1923. Thanks again!
